### PR TITLE
feat(merge): add testing_commitBlockV1 to commit a block on the chain head

### PR DESCRIPTION
**Pull request by AnkushinDaniil**
## Changes

- Add `testing_commitBlockV1(payloadAttributes, txRlps, extraData?) -> Hash256` to `ITestingRpcModule` and its implementation in `TestingRpcModule`.
- Build the block in a transient `IBlockProducerEnvFactory` scope, suggest it to the main `IBlockTree` with `BlockTreeSuggestOptions.ShouldProcess`, and wait for `NewHeadBlock` keyed on the expected hash before returning (30 s timeout).
- Serialize concurrent invocations behind an internal `SemaphoreSlim` and document the policy in the interface description.
- Implement `IDisposable` on `TestingRpcModule` so Autofac disposes the semaphore on container teardown.
- Thread the new `IBlockTree` constructor parameter through the DI registration in `MergePlugin`.
- Add three behavioral tests for the new endpoint (head commit, null head, suggest failure), ported to master's `IBlockProducerEnvFactory` pattern.
- Drop three unused \`using\` directives flagged by \`Check code lint\`.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

Three new tests cover: success path (block becomes head), \`chain head not found\` failure, and \`failed to commit block\` (suggest returns \`InvalidBlock\`). The existing \`testing_buildBlockV1\` suite continues to pass after the constructor signature change.

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

## Remarks

The branch was opened by @marcindsobczak and @kamil-chodola in December 2025; their original 18 commits are preserved below the merge commit that reconciles 540 commits of master drift. Marcin gave explicit go-ahead to handle the rebase + adaptation; co-author trailers preserve attribution.

\`CancellationToken\` propagation to the inner processor and to \`_commitLock.WaitAsync\` is acknowledged and deferred to a follow-up that touches both endpoints together.

### CODE REVIEW

**Comment by AnkushinDaniil** [comment]
@claude re-review

**Comment by LukaszRozmej** [comment]
@claude review — deduplicated the build/commit pipelines into a shared ProduceBlockAsync (returning ResultWrapper<ProducedBlock>), addressed the review nits (LINQ→foreach in DecodeTransactions, Hash null guard, RLP-decode test for commitBlockV1) and threaded processExitSource.Token through to ChainProcessor.Process for cooperative shutdown.

**Review by LukaszRozmej — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
@claude review — addressed the two remaining nits: SuggestAndWaitForHeadAsync now returns the captured expectedHash instead of re-reading the property, and ProduceBlockAsync distinguishes shutdown-cancelled processing ("node is shutting down") from genuine processing failure.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs [hunk 1]
```diff
@@ -195,46 +198,87 @@ public async Task Returns_error_for_unknown_parent()
...
+    [Test]
+    public async Task Testing_commitBlockV1_commits_block_to_chain_head()
+    {
+        (TestingRpcModule module, IBlockTree blockTree, BlockHeader chainHeadHeader) =
+            CreateCommitTestingModule(suggestResult: AddBlockResult.Added);
+
+        ResultWrapper<Hash256?> result = await module.testing_commitBlockV1(
+            CreateDefaultPayloadAttributes(chainHeadHeader),
+            Array.Empty<byte[]>(),
+            Array.Empty<byte>());
```
Comment (alexb5dh) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs]: `Array.Empty<T>()` can be simplified here and below:
```suggestion
        ResultWrapper<Hash256?> result = await module.testing_commitBlockV1(
            CreateDefaultPayloadAttributes(chainHeadHeader), [], []
        );
```
Unless you want to specify types explicitly.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs [hunk 2]
```diff
@@ -195,46 +198,87 @@ public async Task Returns_error_for_unknown_parent()
         result.Result.Error.Should().Contain("unknown parent block");
     }
 
-    private static (TestingRpcModule module, Hash256 parentHash, BlockHeader parentHeader) CreateDefaultTestingModule(
+    [Test]
+    public async Task Testing_commitBlockV1_commits_block_to_chain_head()
+    {
+        (TestingRpcModule module, IBlockTree blockTree, BlockHeader chainHeadHeader) =
```
Comment (alexb5dh) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs:213]: General recommendation is to dispose  disposable `TestingRpcModule`, even in tests

File: src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs [hunk 3]
```diff
@@ -245,10 +289,84 @@ private static (TestingRpcModule module, Hash256 parentHash, BlockHeader parentH
         IBlockProducerEnvFactory blockProducerEnvFactory = Substitute.For<IBlockProducerEnvFactory>();
         blockProducerEnvFactory.CreateTransient().Returns(new ScopedBlockProducerEnv(blockProducerEnv, Substitute.For<IAsyncDisposable>()));
 
-        TestingRpcModule module = new(blockProducerEnvFactory, gasLimitCalculator, specProvider, blockFinder, LimboLogs.Instance);
+        IBlockFinder blockFinder = Substitute.For<IBlockFinder>();
+        IBlockTree blockTree = Substitute.For<IBlockTree>();
```
Comment (alexb5dh) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs:302]: Would be nice to reuse `TestBlockchain` instead of mocking, but may be out of scope here.

File: src/Nethermind/Nethermind.Merge.Plugin/ITestingRpcModule.cs [hunk 4]
```diff
@@ -18,4 +18,10 @@ public interface ITestingRpcModule : IRpcModule
         IsSharable = true,
         IsImplemented = true)]
     public Task<ResultWrapper<object?>> testing_buildBlockV1(Hash256 parentBlockHash, PayloadAttributes payloadAttributes, IEnumerable<byte[]>? txRlps, byte[]? extraData = null);
```
Comment (alexb5dh) [inline src/Nethermind/Nethermind.Merge.Plugin/ITestingRpcModule.cs]: Can probably change to non-nullable `Task<ResultWrapper<object>>`?

File: src/Nethermind/Nethermind.Merge.Plugin/ITestingRpcModule.cs [hunk 5]
```diff
@@ -18,4 +18,10 @@ public interface ITestingRpcModule : IRpcModule
...
+
+    [JsonRpcMethod(
+        Description = "Build a block from provided transactions on top of the current chain head, commit it, and wait for the BlockchainProcessor to advance the canonical head to the committed block before returning. Returns the committed block hash. Concurrent invocations are serialized inside the implementation, so callers do not need their own mutex; serialization order between simultaneous calls is unspecified.",
+        IsSharable = true,
+        IsImplemented = true)]
+    public Task<ResultWrapper<Hash256?>> testing_commitBlockV1(PayloadAttributes payloadAttributes, IEnumerable<byte[]> txRlps, byte[]? extraData = null);
```
Comment (alexb5dh) [inline src/Nethermind/Nethermind.Merge.Plugin/ITestingRpcModule.cs]: Same here

File: src/Nethermind/Nethermind.Merge.Plugin/TestingRpcModule.cs [hunk 6]
```diff
@@ -29,66 +33,162 @@ public class TestingRpcModule(
...
         {
-            IReleaseSpec spec = specProvider.GetSpec(new ForkActivation(parentBlock.Header.Number + 1, payloadAttributes.Timestamp));
+            return ResultWrapper<Hash256?>.Fail("node is shutting down", ErrorCodes.InternalError);
+        }
 
-            BlockHeader header = PrepareBlockHeader(parentBlock.Header, payloadAttributes, spec, extraData);
+        try
+        {
+            BlockHeader? chainHead = blockTree.Head?.Header;
+            if (chainHead is null)
```
Comment (alexb5dh) [inline src/Nethermind/Nethermind.Merge.Plugin/TestingRpcModule.cs]: Subjective, but:
```suggestion
            if (blockTree.Head?.Header is not BlockHeader chainHead)
```

File: src/Nethermind/Nethermind.Merge.Plugin/TestingRpcModule.cs [hunk 7]
```diff
@@ -29,66 +33,162 @@ public class TestingRpcModule(
...
+        return ResultWrapper<object?>.Success(CreateGetPayloadResult(produced.Data.Block, produced.Data.Fees, produced.Data.Spec));
+    }
+
+    public async Task<ResultWrapper<Hash256?>> testing_commitBlockV1(
+        PayloadAttributes payloadAttributes, IEnumerable<byte[]> txRlps, byte[]? extraData = null)
+    {
+        CancellationToken exitToken = processExitSource.Token;
+        try
+        {
+            await _commitLock.WaitAsync(exitToken).ConfigureAwait(false);
```
Comment (alexb5dh) [inline src/Nethermind/Nethermind.Merge.Plugin/TestingRpcModule.cs]: Don't think we need `ConfigureAwait` anywhere - it's not a library/package

**Review by asdacap — APPROVED** [review]

**Review by damian-orzechowski — APPROVED** [review]

**Review by marcindsobczak — APPROVED** [review]

File: src/Nethermind/Nethermind.Merge.Plugin/TestingRpcModule.cs [hunk 8]
```diff
@@ -91,6 +99,112 @@ public class TestingRpcModule(
...
+            if (processedBlock.Transactions.Length != transactions.Length)
+            {
+                string error = $"expected {transactions.Length} transactions but only {processedBlock.Transactions.Length} were included";
+                if (_logger.IsWarn) _logger.Warn($"testing_commitBlockV1 failed: {error}");
+                return ResultWrapper<Hash256?>.Fail(error, ErrorCodes.InvalidInput);
+            }
+
+            // SuggestBlockAsync only enqueues; subscribe to NewHeadBlock before submitting
+            // so we don't miss the head-advance signal on fast paths.
+            Hash256 expectedHash = processedBlock.Hash!;
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Merge.Plugin/TestingRpcModule.cs]: Addressed in the merged version of this PR — `processedBlock.Hash is null` check returns `ResultWrapper<Hash256>.Fail("processed block has no hash", ErrorCodes.InternalError)` before the read. Verified in current master.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs (hunk 1 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs]: Addressed in the merged version — the test file was reorganized around `TestBlockchain` (per the related thread r3157856622), so all `Array.Empty<T>()` call sites went away with the rewrite. Verified: `grep Array.Empty` returns zero in master's TestingRpcModuleTests.cs.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs (hunk 2 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs:213]: Addressed via the larger refactor in the merged version — tests now use `TestBlockchain` instead of constructing `TestingRpcModule` directly, so the disposal lifecycle is owned by the harness. Verified in current master.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs (hunk 3 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/TestingRpcModuleTests.cs:302]: Done in the merged version — tests are now built on `TestBlockchain` (replacing the NSubstitute mocks in this PR). Verified: `grep -c using.module` and `grep -c Array.Empty` both go to zero in master's test file because the structure is entirely different.

File: src/Nethermind/Nethermind.Merge.Plugin/ITestingRpcModule.cs (hunk 4 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Merge.Plugin/ITestingRpcModule.cs]: Addressed in the merged version — the interface is now `Task<ResultWrapper<object>> testing_buildBlockV1(...)` (and `Task<ResultWrapper<Hash256>> testing_commitBlockV1(...)`, even tighter than originally suggested). Verified in current master.
